src/language-js/parser-babel.js

```diff
@@ -65,6 +65,10 @@ function createParse(parseMethod, ...pluginCombinations) {
     // Inline the require to avoid loading all the JS if we don't use it
     const babel = require("@babel/parser");
 
+    // HTML-like Comments https://tc39.es/ecma262/#sec-html-like-comments
+    // requires sourceType to be `script` https://github.com/babel/babel/issues/7802#issuecomment-492522145
+    const sourceType = /<!--|-->/.test(text) ? "unambiguous" : "module";
```

Comment:
This regular expression doesn't do the thing you want; e.g. it would match

```js
let x=10;while(x-->0)console.log(x)
await(import('mod'))
```
which does not actually have contain HTML-like comment and which has a substantially different AST when interpreted as a module vs as a script.

Reply:
This is a  real good case,  HTML-like comment and dynamic import , I'll add to tests.

This also why I set it to `unambiguous` instead of `script`, the output seems same as stable version.

[stable](https://prettier.io/playground/?N4Igxg9gdgLgprEAuEBDJsAWBaANqgZxmzEwEtcATACgCYoACAagYEYBKAHUYYwhmrcGwhgDoCEALZwS+AgSEjxUmf0xwATrMIKew5dOwA3TQE88ES1es2rUAObb5i4aly5FXKC3RY8hYlIKGnpmNi9hPgEXMQlDMDldEViVbDVNJySlOJkTDXNcWyKbB0yYtw8eCN4kACM4ADMIDThgAF8Gbi6oUXQ4AA94KBpRMigABwBXGCRKMgJUWtw4SnZQABoQCHGYMmgCZFBUDQ0IAHcABWOEA5Q3M9RTA83ajVQwAGs4GABlcfexvZkDANJM4JtMDBJLgAOrkeAEf5gOA-G5kXZGdGmZDgeQgTZjAiaGAXN72SSoZANNxEzYAKwI-QAQm9Pt8fqhpAAZMZwKk08EgBn9H6A5YARUm-D5SGpuFpIH+GiJGhxtUWcFw+MVGjGMBhZEoMEwyAAHAAGTbjU5EmFvcY461wFUmbUARyl8FJ21uaAI2CgcBWK21LQ9ZBapNQ5MpsoFmyJkjIwNBgoIYrgkul-PlgpgiwNRpNSFomxBqAogIAwlIKTjnQBWbWTIkAFUWtzlCqMYIAksMEL8wLqdgBBYY-GCmZY5oltNpAA#N4Igxg9gdgLgprEAuEAbOMAEAPAvARgAYBuAdwAsBLdACmwFp6A+QgSkigGcJ0A6VCAHM6rADpQAhqQmUYNSgFsADhABOcgOQKIAEw2tWIADQgISmJWidkoCatURSABTsJrKCamkBPayYBGqhJgANYYAMpKwZRQgsgwqgCucCbkMAqoAOpU8JxRYHDhbrKUAG6y3sjgnH4gMZxw6k5BggoSyABmng0mAFac2ABCQaEREgpwADIxcJ3dKSD92OExgugAiokQ8HOoPSBRqg2qVf4S-nCoxgeqMTCZlDow5MgAHIQmSg4NmUFKVV84MdSrMTABHLbwZpmdwgCScehQOBwHQo66qOAQygY5oSVrtJBdPYLBoKSjxJIk1YbSGzQnzEwwc4PJ4vJAAJkZQWoqwAwhAFG0qkCAKzXRINAAq53cRP2pWSAEkoKjYOEwLdzABBFXhGDedC7BoAX2NQA) vs [pr](https://prettier.io/playground/?N4Igxg9gdgLgprEAuEBDJsAWBaANqgZxmzEwEtcATACgCYoACAagYEYBKAHUYYwhmrcGwhgDoCEALZwS+AgSEjxUmf0xwATrMIKew5dOwA3TQE88ES1es2rUAObb5i4aly5FXKC3RY8hYlIKGnpmNi9hPgEXMQlDMDldEViVbDVNJySlOJkTDXNcWyKbB0yYtw8eCN4kACM4ADMIDThgAF8Gbi6oUXQ4AA94KBpRMigABwBXGCRKMgJUWtw4SnZQABoQCHGYMmgCZFBUDQ0IAHcABWOEA5Q3M9RTA83ajVQwAGs4GABlcfexvZkDANJM4JtMDBJLgAOrkeAEf5gOA-G5kXZGdGmZDgeQgTZjAiaGAXN72SSoZANNxEzYAKwI-QAQm9Pt8fqhpAAZMZwKk08EgBn9H6A5YARUm-D5SGpuFpIH+GiJGhxtUWcFw+MVGjGMBhZEoMEwyAAHAAGTbjU5EmFvcY461wFUmbUARyl8FJ21uaAI2CgcBWK21LQ9ZBapNQ5MpsoFmyJkjIwNBgoIYrgkul-PlgpgiwNRpNSFomxBqAogIAwlIKTjnQBWbWTIkAFUWtzlCqMYIAksMEL8wLqdgBBYY-GCmZY5oltNpAA#N4Igxg9gdgLgprEAuEAbOMAEAPAvARgAYBuAdwAsBLdACmwFp6A+QgSkigGcJ0A6VCAHM6rADpQAhqQmUYNSgFsADhABOcgOQKIAEw2tWIADQgISmJWidkoCatURSABTsJrKCamkBPayYBGqhJgANYYAMpKwZRQgsgwqgCucCbkMAqoAOpU8JxRYHDhbrKUAG6y3sjgnH4gMZxw6k5BggoSyABmng0mAFac2ABCQaEREgpwADIxcJ3dKSD92OExgugAiokQ8HOoPSBRqg2qVf4S-nCoxgeqMTCZlDow5MgAHIQmSg4NmUFKVV84MdSrMTABHLbwZpmdwgCScehQOBwHQo66qOAQygY5oSVrtJBdPYLBoKSjxJIk1YbSGzQnzEwwc4PJ4vJAAJkZQWoqwAwhAFG0qkCAKzXRINAAq53cRP2pWSAEkoKjYOEwLdzABBFXhGDedC7BoAX2NQA)

Reply:
That's surprising, because when parsed as a script this should be a function call, not an actual `await`.

Reply:
yes, I  know that , but your case should be a  seperate issue.

Reply:
By the way, do you know any other syntax requires `script`?

Reply:
The only cases where there's something which is valid in both a script and a module but with a different interpretation are HTML comments and `await`. There's a few things which are legal in scripts but not modules, like octal literals (`x = 01` is legal as a script but not a module), and of course import/export are legal in modules but not scripts.

The only interesting case with HTML comments when you have `<!--` in a module, which doesn't really come up. e.g.

```js
let x = 2;
if (-1<!--x) console.log(1); if(0
==-1) console.log(2);
```
prints 1 in a script and 2 in a module - but who writes that?